sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala

```diff
@@ -0,0 +1,125 @@
+        // back to non-adaptive mode if adaptive execution is supported in any of the sub-queries.
+        logDebug(s"Adaptive execution enabled for plan: $plan")
+        val subqueryMap = buildSubqueryMap(plan)
+        AdaptiveSparkPlanExec(plan, session, subqueryMap, stageCache)
+      } catch {
+        case SubqueryAdaptiveNotSupportedException(subquery) =>
+          logWarning(s"${SQLConf.RUNTIME_REOPTIMIZATION_ENABLED.key} is enabled " +
+            s"but is not supported for sub-query: $subquery.")
+          plan
+      }
+    case _ =>
+      if (conf.runtimeReoptimizationEnabled) {
+        logWarning(s"${SQLConf.RUNTIME_REOPTIMIZATION_ENABLED.key} is enabled " +
+          s"but is not supported for query: $plan.")
+      }
+      plan
+  }
+
+  private def supportAdaptive(plan: SparkPlan): Boolean = {
+    sanityCheck(plan) &&
+      !plan.logicalLink.exists(_.isStreaming) &&
+      plan.children.forall(supportAdaptive)
+  }
+
+  private def sanityCheck(plan: SparkPlan): Boolean =
+    plan.logicalLink.isDefined
+
+  /**
+   * Returns an expression-id-to-execution-plan map for all the sub-queries.
+   * For each sub-query, generate the adaptive execution plan for each sub-query by applying this
+   * rule, or reuse the execution plan from another sub-query of the same semantics if possible.
+   */
+  private def buildSubqueryMap(plan: SparkPlan): Map[Long, ExecSubqueryExpression] = {
+    val subqueryMapBuilder = mutable.HashMap.empty[Long, ExecSubqueryExpression]
+    plan.foreach(_.expressions.foreach(_.foreach {
+      case expressions.ScalarSubquery(p, _, exprId)
+          if !subqueryMapBuilder.contains(exprId.id) =>
+        val executedPlan = getExecutedPlan(p)
+        val scalarSubquery = execution.ScalarSubquery(
+          SubqueryExec(s"subquery${exprId.id}", executedPlan), exprId)
+        subqueryMapBuilder.put(exprId.id, scalarSubquery)
+      case _ =>
+    }))
+
+    // Reuse subqueries
+    if (session.sessionState.conf.subqueryReuseEnabled) {
+      // Build a hash map using schema of subqueries to avoid O(N*N) sameResult calls.
+      val reuseMap = mutable.HashMap[StructType, mutable.ArrayBuffer[BaseSubqueryExec]]()
+      subqueryMapBuilder.keySet.foreach { exprId =>
+        val sub = subqueryMapBuilder(exprId)
+        val sameSchema =
+          reuseMap.getOrElseUpdate(sub.plan.schema, mutable.ArrayBuffer.empty)
+        val sameResult = sameSchema.find(_.sameResult(sub.plan))
+        if (sameResult.isDefined) {
+          val newExpr = sub.withNewPlan(ReusedSubqueryExec(sameResult.get))
+          subqueryMapBuilder.update(exprId, newExpr)
+        } else {
+          sameSchema += sub.plan
+        }
+      }
+    }
+
+    subqueryMapBuilder.toMap
+  }
+
+  private def getExecutedPlan(plan: LogicalPlan): SparkPlan = {
+    val queryExec = new QueryExecution(session, plan)
+    val adaptivePlan = this.apply(queryExec.sparkPlan)
```

Comment:
Let's add a note here: we apply this rule on the subqueries so that the `AdapativeSparkPlanExec` for subqueries share the same `stageCache` with main query's `AdapativeSparkPlanExec`